### PR TITLE
Add getMoveGroupClient() to move_group_interface

### PR DESCRIPTION
**Pull request by mintar**
Use case: I want to execute a trajectory asynchronously while also being able to monitor its execution and perhaps abort it.

* `MoveGroupInterface::asyncMove()` lets me start asynchronous execution
* `MoveGroupInterface::stop()` lets me abort it
* but there is no way to monitor the execution or get the result

This PR simply adds a getter for the internal `move_action_client`. This enables code like shown in the usage example below. I'm not sure that exposing the internal `move_action_client` is the nicest API (comments welcome), but at least it seems more elegant than replicating all of the client's API one-by-one. In the example below, I've only used `getState()`, `getResult()` and `cancelGoal()`; but the other methods are useful as well (for example `waitForResult()`).

### usage example

```cpp
moveit::planning_interface::MoveGroupInterface group;
// [...]
group.setNamedTarget("home");
group.plan(plan);
group.asyncMove();
std::shared_ptr<actionlib::SimpleActionClient<moveit_msgs::MoveGroupAction> >
    move_action_client = group.getMoveActionClient();

ros::Rate rate(10);
while (!move_action_client->getState().isDone() && ros::ok()) {
  ROS_DEBUG("move_action_client: %s", move_action_client->getState().toString().c_str());
  rate.sleep();
  // to abort execution of the trajectory: `move_action_client->cancelGoal();` or `group.stop();`
}

ROS_DEBUG("move_action_client: %s %s",
          move_action_client->getState().toString().c_str(),
          move_action_client->getState().getText().c_str());

auto error_code = move_action_client->getResult()->error_code.val;

if (error_code == moveit::planning_interface::MoveItErrorCode::SUCCESS) {
  ROS_INFO("Moving to home pose SUCCESSFUL");
  return 0;
} else {
  ROS_ERROR("Moving to home pose FAILED");
  return 1;
}
```

### TODOs

- [x] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [ ] Extended the tutorials / documentation, if necessary [reference](http://moveit.ros.org/documentation/contributing/)
- [ ] Created tests, which fail without this PR [reference](http://docs.ros.org/kinetic/api/moveit_tutorials/html/doc/tests.html)
- [ ] Decide if this should be cherry-picked to other current ROS branches
- [ ] While waiting for someone to review your request, please consider reviewing [another open pull request](https://github.com/ros-planning/moveit/pulls) to support the maintainers. You can refer to our [ROS code review guide](https://github.com/rosin-project/ros_code_review_guide/blob/master/README.md) for direction

### CODE REVIEW

**Comment by v4hn** [comment]
As discussed in private emails, I support this idea.
The MGI is supposed to make the users life *easier* and should not prevent useful applications,
such as aborting actions, looking at feedback or results in asyncMove.

I don't like the idea of sharing the action client though and would prefer to return a reference from the method.
Why did you decide to use a shared_ptr instead?

For consistency it might also make sense to change the method name to `getMoveGroupClient`.

Do you have to write out the pointer definition, or is there a typedef you can use?

**Comment by mintar** [comment]
Regarding the return type:  I could return a reference to the underlying object directly (so, a `const actionlib::SimpleActionClient<moveit_msgs::MoveGroupAction>&`). Now that I think about it, that sounds even better.

File: moveit_ros/planning_interface/move_group_interface/include/moveit/move_group_interface/move_group_interface.h [hunk 1]
```diff
@@ -48,7 +48,9 @@
 #include <moveit_msgs/Grasp.h>
 #include <moveit_msgs/PlaceLocation.h>
 #include <moveit_msgs/MotionPlanRequest.h>
+#include <moveit_msgs/MoveGroupAction.h>
 #include <geometry_msgs/PoseStamped.h>
+#include <actionlib/client/simple_action_client.h>
```
Comment (v4hn) [inline moveit_ros/planning_interface/move_group_interface/include/moveit/move_group_interface/move_group_interface.h]: This line makes `actionlib` a `CATKIN_DEPENDS` of the `moveit_ros_planning_interface` package,
i.e. packages including the header have to be able to find actionlib headers.

Can you please add that to the `CMakeLists.txt`?

File: moveit_ros/planning_interface/move_group_interface/include/moveit/move_group_interface/move_group_interface.h [hunk 2]
```diff
@@ -683,6 +685,11 @@ class MoveGroupInterface
       This call is not blocking (does not wait for the execution of the trajectory to complete). */
   MoveItErrorCode asyncMove();
 
+  /** \brief Get the move_group action client used by the \e MoveGroupInterface.
+      The client can be used for querying the execution state of the trajectory and abort trajectory execution
+      during asynchronous execution. */
+  const actionlib::SimpleActionClient<moveit_msgs::MoveGroupAction>& getMoveGroupClient();
```
Comment (v4hn) [inline moveit_ros/planning_interface/move_group_interface/include/moveit/move_group_interface/move_group_interface.h]: can you actually abort trajectories with a `const` client?
I would have expected this has to be non-const.

**Comment by mintar** [comment]
Ok, all requested changes are done.

File: moveit_ros/planning_interface/move_group_interface/include/moveit/move_group_interface/move_group_interface.h (hunk 1 above)
Comment (mintar) [reply moveit_ros/planning_interface/move_group_interface/include/moveit/move_group_interface/move_group_interface.h]: Done.

File: moveit_ros/planning_interface/move_group_interface/include/moveit/move_group_interface/move_group_interface.h (hunk 2 above)
Comment (mintar) [reply moveit_ros/planning_interface/move_group_interface/include/moveit/move_group_interface/move_group_interface.h]: Yep, you're right, very good catch! `cancelGoal()` was commented out in my test code, so the compiler didn't catch that. Resolved.

**Comment by v4hn** [comment]
Looks good to me. Thanks for your time @mintar :+1: 
Let's see what someone else has to say.

File: moveit_ros/planning_interface/move_group_interface/src/move_group_interface.cpp [hunk 3]
```diff
@@ -1502,6 +1505,12 @@ moveit::planning_interface::MoveItErrorCode moveit::planning_interface::MoveGrou
   return impl_->move(false);
 }
 
+actionlib::SimpleActionClient<moveit_msgs::MoveGroupAction>&
+moveit::planning_interface::MoveGroupInterface::getMoveGroupClient()
```
Comment (rhaschke) [inline moveit_ros/planning_interface/move_group_interface/src/move_group_interface.cpp]: This method is missing the const  specifier, while the implementation method has one.
```suggestion
moveit::planning_interface::MoveGroupInterface::getMoveGroupClient() const
```

Comment (mintar) [reply moveit_ros/planning_interface/move_group_interface/src/move_group_interface.cpp]: Good catch, I obviously meant to make it const everywhere. Fixed!

**Comment by mintar** [comment]
Nice! That went faster than expected.

**Comment by v4hn** [comment]
@mintar Glad you talked to me about it :+1:

**Comment by davidyanghit** [comment]
hi, I discovered there is a bug in current` getMoveGroupClient()`. To get the execution state of `aysncExecute()`, the return client should be `return * execute_action_client_ `instead of `*move_action_client_`. The latter one only works when you use `move()` and `plan(`), while  `execute_action_client_` will always work in `aysncExecute().`

**Comment by simonschmeisser** [comment]
@davidyanghit it would be really nice if you could open a new PR with your proposed changes, it's easier to understand and comment then

**Pull request merged**